File: kata/4kyu/sum_of_perfect_squares/imp.go
package kata

var dp = make([]uint64, 1)

func SumOfSquares(n uint64) uint64 {
	if n < uint64(len(dp)-1) {
		return dp[n]
	}

	start := uint64(len(dp) - 1)
	end := n + 1

	for i := start; i <= end; i++ {
		dp = append(dp, n)
		for j := uint64(1); j*j <= i; j++ {
			dp[i] = min(dp[i], dp[i-j*j]+1)
		}
	}

	return dp[n]
}

File: kata/4kyu/sum_of_perfect_squares/imp.js
var dp = [0]

function sumOfSquares(n) {
    if (n < dp.length) {
        return dp[n];
    }

    let start = dp.length;
    let end = n + 1;

    for (let i = start; i <= end; i++) {
        dp[i] = n + 1;
        for (let j = 1; j * j <= i; j++) {
            dp[i] = Math.min(dp[i], dp[i - j * j] + 1);
        }
    }

    return dp[n];
}

console.log(sumOfSquares(13)); // 2
console.log(sumOfSquares(12)); // 3

